lib/dal/src/workspace_snapshot/split_snapshot.rs

```diff
@@ -0,0 +1,1634 @@
+    }
+
+    #[instrument(name = "split_snapshot.find", level = "debug", skip_all, fields())]
+    pub async fn find(
+        ctx: &DalContext,
+        split_snapshot_supergraph_addr: WorkspaceSnapshotAddress,
+    ) -> WorkspaceSnapshotResult<Self> {
+        let snapshot = match ctx
+            .layer_db()
+            .split_snapshot_supergraph()
+            .read_wait_for_memory(&split_snapshot_supergraph_addr)
+            .await
+        {
+            Ok(supergraph) => {
+                let supergraph = supergraph.ok_or(
+                    WorkspaceSnapshotError::SplitSnapshotSuperGraphMissingAtAddress(
+                        split_snapshot_supergraph_addr,
+                    ),
+                )?;
+
+                let mut subgraphs = vec![];
+                for &subgraph_address in supergraph.addresses() {
+                    let subgraph_address = subgraph_address.into();
+                    let subgraph = ctx
+                        .layer_db()
+                        .split_snapshot_subgraph()
+                        .read_wait_for_memory(&subgraph_address)
+                        .await?
+                        .ok_or(
+                            WorkspaceSnapshotError::SplitSnapshotSubGraphMissingAtAddress(
+                                subgraph_address,
+                            ),
+                        )?;
+
+                    // xxx: we have to make the splitgraph constructable from arcs, it will
+                    // xxx: need to handle the copy-on-write behavior internally
+                    subgraphs.push(subgraph.as_ref().clone());
+                }
+
+                Arc::new(SplitSnapshotGraph::V1(
+                    SplitSnapshotGraphVCurrent::from_parts(supergraph.as_ref().clone(), subgraphs),
+                ))
+            }
+            Err(err) => match err {
+                LayerDbError::Postcard(_) => {
+                    return Err(WorkspaceSnapshotError::WorkspaceSnapshotNotMigrated(
+                        split_snapshot_supergraph_addr,
+                    ));
+                }
+                err => Err(err)?,
+            },
+        };
+
+        Ok(Self {
+            address: Arc::new(Mutex::new(split_snapshot_supergraph_addr)),
+            read_only_graph: snapshot,
+            working_copy: Arc::new(RwLock::new(None)),
+            cycle_check: Arc::new(AtomicBool::new(false)),
+            dvu_roots: Arc::new(Mutex::new(HashSet::new())),
+            inferred_connection_graph: Arc::new(RwLock::new(None)),
+        })
+    }
+
+    pub async fn find_for_change_set(
+        ctx: &DalContext,
+        change_set_id: ChangeSetId,
+    ) -> WorkspaceSnapshotResult<Self> {
+        // There's a race between finding which address to retrieve and actually retrieving it
+        // where it's possible for the content at the address to be garbage collected, and no
+        // longer be retrievable. We'll re-fetch which snapshot address to use, and will retry,
```

Comment:
How often does this happen? Should we trace this at the warn level?

Reply:
you know, i copied this from the existing find code. I don't know if the race is still a problem